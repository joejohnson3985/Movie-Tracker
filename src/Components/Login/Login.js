import React, { Component } from "react";
import { fetchUserData } from "../../APICalls/APICalls";
import { userLogin, setFavorites, isLoggedIn } from "../../Actions";
import "./Login.scss";
import { connect } from "react-redux";

class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: "",
      password: "",
      status: ""
    };
  }

  handleChange = e => {
    this.setState({
      [e.target.type]: e.target.value.toLowerCase()
    });
  };

  handleSubmit = e => {
    e.preventDefault();
    const url = "http://localhost:3000/api/users";
    const userInput = this.state;
    const userOptionObject = {
      method: "POST",
      body: JSON.stringify(userInput),
      headers: {
        "Content-Type": "application/json"
      }
    };
    fetchUserData(url, userOptionObject)
      .then(results => this.props.addCurrentUser(results.data))
      .then(results => this.getFavoriteMovies(results.user.id))
      .catch(error => alert("Email or Password is incorrect"));
  };

  getFavoriteMovies = id => {
    const url = `http://localhost:3000/api/users/${id}/favorites`;
    fetchUserData(url)
      .then(response => this.props.setFavorites(response.data))
      .then(response => this.props.isLoggedIn(true))
      .catch(error => console.log(error));
  };

  render() {
    return (
      <div className="btn-container">
        <form onSubmit={this.handleSubmit} className="login-form">
          <input
            className="email"
            onChange={this.handleChange}
            type="email"
            placeholder="Email"
            value={this.state.email}
          />
          <input
            className="password"
            onChange={this.handleChange}
            type="password"
            placeholder="Password"
            value={this.state.password}
          />
          <button> Sign in </button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  user: state.currentUser
});

const mapDispatchToProps = dispatch => ({
  addCurrentUser: user => dispatch(userLogin(user)),
  setFavorites: favorites => dispatch(setFavorites(favorites)),
  isLoggedIn: bool => dispatch(isLoggedIn(bool))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Login);
