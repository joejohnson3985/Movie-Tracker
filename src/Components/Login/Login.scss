@import "../../Utilities/_Variables.scss";

.login-form {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  width: 425px;
  margin-top: 25px;
  grid-template-areas:
    "email password"
    "button button";

  input {
    padding: 5px;
    margin: 5px 10px;
  }

  button {
    padding: 5px;
    margin: 5px 10px;
    grid-area: button;
  }
}

.email {
  grid-area: email;
}

.password {
  grid-area: password;
}

.btn-container {
  height: 85vh;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  background-image: url("../../Images/main-bg.png");
  background-size: cover;
  background-position: center;
}
