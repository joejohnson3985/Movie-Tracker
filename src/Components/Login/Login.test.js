import {Login}  from './Login.js'
import { shallow, mount } from 'enzyme';
import React from 'react';

window.alert = jest.fn();
let wrapper;
let mockFunction = jest.fn()
let fetchUserData = jest.fn()

describe('Login', () => {

  beforeEach(() => {
    wrapper = shallow(<Login />)
    fetch = fetchUserData.mockImplementation(() => Promise.resolve({ok: true, status: 200,
      json: () => Promise.resolve(mockData),
    }));
  })

  it('Should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Should have a defualt state', () => {
    let expectedState = {email: '', password: '', status: ''}
    expect(wrapper.state()).toEqual(expectedState)
  })

  it('should update state when typing', () => {
    expect(wrapper.state().email).toEqual('');
    let mockEvent = {target:{value:'123@email.com', type: 'email'}}
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state().email).toEqual('123@email.com');
  })

  it('Should fetch post user data to server when logging in', () => {
    let url ="http://localhost:3000/api/users";
    let mockEvent = {preventDefault: jest.fn()}
    wrapper.instance().handleSubmit(mockEvent);
    expect(fetch).toBeCalledWith(url,{"body": "{\"email\":\"\",\"password\":\"\",\"status\":\"\"}", "headers": {"Content-Type": "application/json"}, "method": "POST"})
  })

  it.skip('Should fetch users favorites', () => {
    let id = 1
    let url = 'http://localhost:3000/api/users/1/favorites';
    wrapper.instance().getFavoriteMovies(id)
    expect(fetch).toHaveBeenCalledWith(url) 
  })
})