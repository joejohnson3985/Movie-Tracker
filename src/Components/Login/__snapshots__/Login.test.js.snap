// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Should match the snapshot 1`] = `ShallowWrapper {}`;
